contracts/Minter.vy

```diff
@@ -32,13 +53,85 @@ minted: public(HashMap[address, HashMap[address, uint256]])
 # minter -> user -> can mint?
 allowed_to_mint_for: public(HashMap[address, HashMap[address, bool]])
 
+emergency_return: public(address)
+admin: public(address)
 
 @external
-def __init__(_token: address, _controller: address):
+def __init__(_token: address, _controller: address, _emergency_return: address, _admin: address):
     self.token = _token
     self.controller = _controller
+    self.emergency_return = _emergency_return
+    self.admin = _admin
+
+    self.start_epoch_time = block.timestamp + INFLATION_DELAY - RATE_REDUCTION_TIME
+    self.mining_epoch = -1
+    self.rate = 0
+    self.is_start = True
+    self.committed_rate = MAX_UINT256
 
 
+@internal
+def _update_mining_parameters():
+    """
+    @dev Update mining rate and supply at the start of the epoch
+         Any modifying mining call must also call this
+    """
+    _rate: uint256 = self.rate
+
+    self.start_epoch_time += RATE_REDUCTION_TIME
+    self.mining_epoch += 1
+
+    if _rate == 0 and self.is_start:
+        _rate = INITIAL_RATE
+        self.is_start = False
+    else:
+        _committed_rate: uint256 = self.committed_rate
+        if _committed_rate != MAX_UINT256:
+          _rate = _committed_rate
+          self.committed_rate = MAX_UINT256
+
+    self.rate = _rate
+
+    log UpdateMiningParameters(block.timestamp, _rate)
+
+@external
+def update_mining_parameters():
```

Comment:
update_mining_parameters(), start_epoch_time_write(), future_epoch_time_write() identical as in [ERC20CRV](https://github.com/curvefi/curve-dao-contracts/blob/4e428823c8ae9c0f8a669d796006fade11edb141/contracts/ERC20CRV.vy#L125)